geopandas/base.py

```diff
@@ -1202,6 +1203,35 @@ def explode(self):
         dtype: geometry
 
         """
+
+        # TODO: use version check instead of hasattr below
+        # if compat.HAS_PYGEOS and compat.PYGEOS_GE_09:
+        if compat.HAS_PYGEOS:
+            import pygeos  # noqa
+
+            if hasattr(pygeos, "get_parts"):
+
+                geometries, outer_index = pygeos.get_parts(
+                    self.values.data, return_index=True
+                )
+
+                if len(outer_index):
+                    # generate inner index
+                    _, counts = np.unique(outer_index, return_counts=True)
+                    inner_index = np.concatenate([np.arange(c) for c in counts])
```

Comment:
I was looking at the pandas cumcount implementation, and they are basically doing this:

```
ids = outer_index

run = np.r_[True, ids[:-1] != ids[1:]]
rep = np.diff(np.r_[np.nonzero(run)[0], len(ids)])
out = (~run).cumsum()
out -= np.repeat(out[run], rep)

inner_index = out
```

I didn't yet do the effort to fully understand those lines of code ... but, it seems to result in the same, and be considerably faster (especially in case with only few exploded geometries).

```
In [8]: df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
In [10]: df = pd.concat([df]*100, axis=0)

In [11]: df = df.reset_index(drop=True)

In [12]: %timeit pygeos.get_parts(df.geometry.array.data, return_index=True)
21.9 ms ± 331 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [13]: _, outer_index = pygeos.get_parts(df.geometry.array.data, return_index=True)

In [14]: %%timeit
    ...: _, counts = np.unique(outer_index, return_counts=True)
    ...: np.concatenate([np.arange(c) for c in counts])
24.7 ms ± 521 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [15]: ids = outer_index

In [16]: %%timeit
    ...: run = np.r_[True, ids[:-1] != ids[1:]]
    ...: rep = np.diff(np.r_[np.nonzero(run)[0], len(ids)])
    ...: out = (~run).cumsum()
    ...: out -= np.repeat(out[run], rep)
    ...: 
    ...: 
263 µs ± 10.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Reply:
Nice!  I knew there had to be a better way to do this with numpy but couldn't find it.